js/packages/seed-vault/android/src/main/java/com/solanamobile/seedvault/reactnative/SolanaMobileSeedVaultLibModule.kt

```diff
@@ -0,0 +1,392 @@
+            pushString(derivationPath)
+        })
+    }
+
+    @ReactMethod
+    fun requestPublicKeys(authToken: String, derivationPaths: ReadableArray) {
+        Log.d(TAG, "Requesting public keys for provided derviation paths...")
+        val intent = Wallet.requestPublicKeys(authToken.toLong(), Arguments.toList(derivationPaths)?.mapNotNull {
+            (it as? String)?.let { uriString -> Uri.parse(uriString) }
+        } as ArrayList ?: arrayListOf())
+        reactContext.currentActivity?.startActivityForResult(intent, REQUEST_GET_PUBLIC_KEYS);
+    }
+
+    private fun sendEvent(reactContext: ReactContext, eventName: String, params: WritableMap? = null) {
+        reactContext
+            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
+            .emit(eventName, params)
+    }
+
+    private fun sendSeedVaultEvent(event: SeedVaultEvent) {
+        sendEvent(reactContext, Companion.SEED_VAULT_EVENT_BRIDGE_NAME, event.toWritableMap())
+    }
+
+    private fun observeSeedVaultContentChanges() {
+        reactContext.contentResolver.registerContentObserver(
+            WalletContractV1.WALLET_PROVIDER_CONTENT_URI_BASE,
+            true,
+            object : ContentObserver(Handler(reactContext.mainLooper)) {
+                override fun onChange(selfChange: Boolean) =
+                    throw NotImplementedError("Stub for legacy onChange")
+                override fun onChange(selfChange: Boolean, uri: Uri?) =
+                    throw NotImplementedError("Stub for legacy onChange")
+                override fun onChange(selfChange: Boolean, uri: Uri?, flags: Int) =
+                    throw NotImplementedError("Stub for legacy onChange")
+
+                override fun onChange(selfChange: Boolean, uris: Collection<Uri>, flags: Int) {
+                    Log.d(TAG, "Received change notification for $uris (flags=$flags); refreshing viewmodel")
+                    sendEvent(reactContext, SEED_VAULT_CONTENT_CHANGE_EVENT_BRIDGE_NAME, 
+                        params = Arguments.createMap().apply {
+                            putString("__type", "SeedVaultContentChange")
+                            putArray("uris", Arguments.createArray().apply {
+                                uris.forEach { uri ->
+                                    pushString(uri.toString())
+                                }
+                            })
+                        }
+                    )
+                }
+            }
+        )
+    }
+
+    private fun handleActivtyResult(requestCode: Int,  resultCode: Int, data: Intent?) {
+        when (requestCode) {
+            REQUEST_AUTHORIZE_SEED_ACCESS -> {
+                try {
+                    val authToken = Wallet.onAuthorizeSeedResult(resultCode, data)
+                    Log.d(TAG, "Seed authorized, AuthToken=$authToken")
+                    sendSeedVaultEvent(SeedVaultEvent.SeedAuthorized(authToken))
```

Comment:
Is there anything that prevents this from engendering a race condition – ie. these events being fired before the app starts to listen for them?

Some things to consider:

* fataling if there are no listeners
* accumulating events and playing them all into the next listener to attach

Reply:
Yes good call out. Mike and I have been discussing this a lot. 

With this current design, there is nothing stopping the consumer from calling a method like `authorizeNewSeed` _before_ they add a listener. In this case the caller would get nothing back, not even an error. We should at the very least throw some errors if there are no listeners. 

I think we can do even better tho. I think exposing an async-await/promise based API would be much more straightforward for js consumers, and hide the listener under the hood.

Reply:
Given that this library is not yet released on npm, I think we can handle this feedback in a separate PR. I'd like to experiment with wrapping up the native API into an async-await on the js side.